src/software/multithreading/threaded_observer.h

```diff
@@ -64,4 +73,27 @@ class ThreadedObserver : public Observer<T>
     std::thread pull_from_buffer_thread;
 };
 
+template <typename T>
+class FirstInFirstOutThreadedObserver : public GenericThreadedObserver<T>
+{
+   public:
+    FirstInFirstOutThreadedObserver<T>() : GenericThreadedObserver<T>(){};
+    explicit FirstInFirstOutThreadedObserver<T>(size_t buffer_size)
+        : GenericThreadedObserver<T>(buffer_size){};
+    std::optional<T> getNextValue(const Duration& max_wait_time) final;
+};
+
+template <typename T>
+class LastInFirstOutThreadedObserver : public GenericThreadedObserver<T>
+{
+   public:
+    LastInFirstOutThreadedObserver<T>() : GenericThreadedObserver<T>(){};
+    explicit LastInFirstOutThreadedObserver<T>(size_t buffer_size)
+        : GenericThreadedObserver<T>(buffer_size){};
+    std::optional<T> getNextValue(const Duration& max_wait_time) final;
+};
```

Comment:
Could we put these seperate classes in their own files, just to be consistent with the rest of the codebase (generally 1 class per file)?